src/kernel/state/vatState.js

```diff
@@ -0,0 +1,327 @@
+        kernelSlotKey = `${type}-${id}`;
+        break;
+      }
+      case 'export': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.exports;
+        kernelSlotKey = `${type}-${kernelSlot.vatID}-${id}`;
+        break;
+      }
+      case 'device': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.devices;
+        kernelSlotKey = `${type}-${kernelSlot.vatID}-${id}`;
+        break;
+      }
+      default:
+        throw new Error(`unexpected kernelSlot type ${kernelSlot.type}`);
+    }
+    return {
+      typedKernelToVatMap,
+      kernelSlotKey,
+    };
+  }
+
+  function getValSlotTypedMapAndKey(vatSlot) {
+    insistVatSlot(vatSlot);
+    const { type, id } = vatSlot;
+    let typedVatToKernelMap;
+    // imports, deviceImports, promises, resolvers
+
+    switch (type) {
+      case 'import': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.imports;
+        break;
+      }
+      case 'deviceImport': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.deviceImports;
+        break;
+      }
+      case 'promise': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.promises;
+        break;
+      }
+      case 'resolver': {
+        typedVatToKernelMap = state.vatSlotToKernelSlot.resolvers;
+        break;
+      }
+      default:
+        throw new Error(`unexpected vatSlot type ${vatSlot.type}`);
+    }
+    return {
+      typedVatToKernelMap,
+      vatSlotKey: `${type}-${id}`,
+    };
+  }
+
+  function getVatSlotTypeFromKernelSlot(kernelSlot) {
+    switch (kernelSlot.type) {
+      case 'export':
+        return 'import';
+      case 'device':
+        return 'deviceImport';
+      case 'promise':
+        return 'promise';
+      case 'resolver':
+        return 'resolver';
+      default:
+        throw new Error('unrecognized kernelSlot type');
+    }
+  }
+
+  function allocateNextVatSlotID(kernelSlot) {
+    const vatSlotType = getVatSlotTypeFromKernelSlot(kernelSlot);
+
+    insist(
+      allowedVatSlotTypes.includes(vatSlotType),
+      `type ${vatSlotType} is not an allowed vat slot type`,
+    );
+
+    const i = state.nextIDs[vatSlotType];
+
+    state.nextIDs[vatSlotType] = i + 1;
+
+    return i;
+  }
+
+  function mapVatSlotToKernelSlot(vatSlot) {
+    const { typedVatToKernelMap, vatSlotKey } = getValSlotTypedMapAndKey(
+      vatSlot,
+    );
+    const kernelSlot = typedVatToKernelMap.get(vatSlotKey);
+    if (kernelSlot === undefined) {
+      throw new Error(`unknown ${vatSlot.type} slot '${vatSlot.id}'`);
+    }
+    return kernelSlot;
+  }
+
+  function mapKernelSlotToVatSlot(kernelSlot) {
+    const { typedKernelToVatMap, kernelSlotKey } = getKernelSlotTypedMapAndKey(
+      kernelSlot,
+    );
+    if (!typedKernelToVatMap.has(kernelSlotKey)) {
+      // must add both directions
+      const newVatSlotID = Nat(allocateNextVatSlotID(kernelSlot));
+      const vatSlotType = getVatSlotTypeFromKernelSlot(kernelSlot);
```

Comment:
it's awkward that `allocateNextVatSlotID` also calls `getVatSlotTypeFromKernelSlot` internally, then throws out the value, and nobody else calls `allocateNextVatSlotID`. Feels like we should be able to merge these somehow. And also that we pass in the whole `kernelSlot` but it really only cares about the type: the API makes it seem like the ID matters too, but it's ignored.